File: jianzhioffer/firstround/jianzhi_55_1.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    # def maxDepth(self, root):  # DFS
    #     """
    #     :type root: TreeNode
    #     :rtype: int
    #     """
    #     if not root:
    #         return 0
    #     return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

    def maxDepth(self, root):  # BFS
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        queue, depth = [root], 0
        while queue:
            tmp = []
            for node in queue:
                if node.left:
                    tmp.append(node.left)
                if node.right:
                    tmp.append(node.right)
            queue = tmp
            depth += 1
        return depth


if __name__ == '__main__':
    # n1 = TreeNode(3)
    # n2 = TreeNode(1)
    # n3 = TreeNode(4)
    # n4 = TreeNode(2)
    #
    # n1.left = n2
    # n1.right = n3
    # n2.right = n4

    n1 = TreeNode(5)
    n2 = TreeNode(3)
    n3 = TreeNode(6)
    n4 = TreeNode(2)
    n5 = TreeNode(4)
    n6 = TreeNode(1)

    n1.left = n2
    n1.right = n3
    n2.left = n4
    n2.right = n5
    n4.left = n6

    obj = Solution()
    number = obj.maxDepth(n1)
    print(number)

File: jianzhioffer/secondround/jianzhi_55_1.cpp
#include <cstddef>
#include <algorithm>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x):val(x), left(NULL), right(NULL) {};
};

// class Solution
// {
// public:
//     int maxDepth(TreeNode* root);

// private:
//     int DFS(TreeNode* root);
// };

// //DFS。时间复杂度O(N)，空间复杂度O(N)(递归占用O(N))
// int Solution::maxDepth(TreeNode* root)
// {
//     if(root==NULL) return 0;
//     return DFS(root);;
// }

// int Solution::DFS(TreeNode* root)
// {
//     if(root==NULL) return 0;
//     return max(DFS(root->left), DFS(root->right)) + 1;
// }

class Solution
{
public:
    int maxDepth(TreeNode* root);

private:
    int BFS(TreeNode* root);
};

//BFS。时间复杂度O(N)，空间复杂度O(N)(递归占用O(N))
int Solution::maxDepth(TreeNode* root)
{
    if(root==NULL) return 0;
    return BFS(root);
}

int Solution::BFS(TreeNode* root)
{
    queue<TreeNode*> q;
    q.push(root);
    int res = 0;
    while (!q.empty())
    {
        res++;
        int size = q.size();
        while (size>0)
        {
            TreeNode* node = q.front();
            q.pop();
            size--;
            if(node->left) q.push(node->left);
            if(node->right) q.push(node->right);
        }
    }
    return res;
}


int main(int argc, char const *argv[])
{
    TreeNode* n1 = new TreeNode(3);
    TreeNode* n2 = new TreeNode(9);
    TreeNode* n3 = new TreeNode(20);
    TreeNode* n4 = new TreeNode(15);
    TreeNode* n5 = new TreeNode(7);

    n1->left = n2;
    n1->right = n3;
    n3->left = n4;
    n3->right = n5;
    Solution obj;
    int result = obj.maxDepth(n1);
    return 0;
}